### PR TITLE
Fix mIoU calculatiton range

### PR DESCRIPTION
**Pull request by sennnnn**
1. Fix the bug about torch.histc function. When setting arg:max same as arg:bins for torch.histc, some channel of mIoU may be nan.
2. Modify tests/test_metrics.py. Add some abnormal num_classes test cases.

# Pascal Context Dataset
Single-scale, No flip, mIoU test result:

| model | num classes | Lr schd | after fix | before fix | readme |
| :--: | :--: | :--: | :--: | :--: | :--: |
| HRNetV2-W48(new) | 59 | 40000 | 50.37 | 50.13 | None |
| HRNetV2-w48(new) | 59 | 80000 | 51.12 | 50.85 | None |
| HRNetV2-W48(new) | 60 | 40000 | 45.55 | 45.55 | 45.14 |
| HRNetV2-W48(old) | 60 | 40000 | 45.14 | 45.14 | 45.14 |

before fix: 
* keyboard=nan

# Cityscapes Dataset

Single-scale, No flip, mIoU test result:

| model | Crop Size | Lr schd | after fix | before fix | readme |
| :--: | :--: | :--: | :--: | :--: | :--: |
| HRNetV2-W48(old) | 512x1024 | 40000 | 78.52 | 78.52 | 78.48 |
| HRNetV2-W48(old) | 512x1024 | 80000 | 79.98 | 79.98 | 79.93 |

# ADE20K Dataset

Single-scale, No flip, mIoU test result:

| model | Crop Size | Lr schd | after fix | before fix | readme |
| :--: | :--: | :--: | :--: | :--: | :--: |
| HRNetV2-W48(old) | 512x512 | 80000 | 41.9 | 41.76 | 41.9 |
| HRNetV2-W48(old) | 512x512 | 160000 | 42.02 | 41.82 | 42.02 |

before fix: 
* fireplace=nan; 
* blind=nan; 
* television receiver=nan;
* botle=nan; 
* cradle=nan; 
* animal=nan;

### CODE REVIEW

**Comment by xvjiarui** [comment]
Could we add some test cases for this?

**Comment by xvjiarui** [comment]
For example, we may construct some test cases that fail in the previous version.

**Review by xiexinch — APPROVED** [review]

**Comment by sennnnn** [comment]
We can use trained model file to test on both two version IoU calculation function. 
The trained model file [BaiduNetDisk(Extract Code: 0l69)](https://pan.baidu.com/s/1-9upD_J0P_fMVt7by1fxqQ)

File: tests/test_metrics.py [hunk 1]
```diff
@@ -60,110 +60,124 @@ def legacy_mean_dice(results, gt_seg_maps, num_classes, ignore_index):
...
-        label,
-        num_classes,
-        ignore_index=255,
-        metrics=['mDice', 'mIoU'],
-        nan_to_num=-1)
-    assert acc[-1] == -1
-    assert dice[-1] == -1
-    assert iou[-1] == -1
+    max_classes = 255
+    for num_classes in range(6, max_classes):
```
Comment (xvjiarui) [inline tests/test_metrics.py:65]: Why testing from 6 to 255?

Comment (sennnnn) [reply tests/test_metrics.py]: I want to expand the test range as wider as possible.

File: tests/test_metrics.py [hunk 2]
```diff
@@ -121,6 +127,17 @@ def test_metrics():
...
+    # Test the bug which is caused by torch.histc.
+    # torch.histc:  https://pytorch.org/docs/stable/generated/torch.histc.html
+    # When the arg:bins is set to be same as arg:max,
+    # some channels of mIoU may be nan.
+    results = np.array([np.repeat(31, 59)])
+    label = np.array([np.arange(59)])
+    num_classes = 59
+    all_acc, acc, iou = eval_metrics(
+        results, label, num_classes, ignore_index=255, metrics='mIoU')
+    assert np.sum(np.isnan(iou)) == 0
```
Comment (xvjiarui) [inline tests/test_metrics.py:139]:
```suggestion
    assert not np.any(np.isnan(iou))
```

**Comment by xvjiarui** [comment]
How does this PR affect mIoU of cityscapes and ADE20k?

**Comment by sennnnn** [comment]
> How does this PR affect mIoU of cityscapes and ADE20k?

mIoU of cityscapes has a little bit wave and doesn't contain nan value. But mIoU of ADE20K has six nan classes.

**Pull request merged**